### DISCUSSION TITLE
Multiprocessing PySWMM simulations

### DISCUSSION THREAD

**Discussion by omarseleem92** [pyswmm/pyswmm#465] [Q&A] [answered]
Hi,

I am currently working on reproducing a research paper titled "Constructing an Inventory for Fast Screening of Urban Green Infrastructure Suitability" (https://iwaponline.com/bgs/article/4/2/213/92196/). The paper presents a simple model comprising two sub-catchments, one representing impervious areas and the other representing green infrastructure.

Initially, I successfully ran the model using PySWMM. However,  I tried to parallelize the simulations using the multiprocessing module to reduce the simulations time. Unfortunately, the parallelization did not work as expected.

I am seeking recommendations or advice on how to effectively run the model in parallel or if there are any potential mistakes in my current implementation.

I have attached the input file (Inp.txt) and the Python scripts for PySWMM (PySWMM.txt) and multiprocessing (PySWMM_multiprocessor.txt).

[Inp.txt](https://github.com/pyswmm/pyswmm/files/12146167/Inp.txt)

[PySWMM.txt](https://github.com/pyswmm/pyswmm/files/12146155/PySWMM.txt)
[PySWMM_mulitprocessor.txt](https://github.com/pyswmm/pyswmm/files/12146156/PySWMM_mulitprocessor.txt)

**Comment by bemcdonnell** [comment]
@omarseleem92, Thanks for reaching out and raising this!  So you are describing a problem that actually has its roots all the way down in SWMM.  SWMM it not thread-safe/re-entrant.  Therefore, per "instance" of Python, you can only open one model at a time.  BUT, there is a workaround using `queue` and `subprocess`.  This workflow essentially runs each model on its own process.  The subprocess command would look something like this:

```
> python pyswmm_wrapper.py <cli-args>
```
For reference, take a look at the pyswmm_wrapper in swmmio (https://github.com/pyswmm/swmmio/blob/master/swmmio/wrapper/pyswmm_wrapper.py)

Once each job is complete, you can return results through a file or json to your main code

**Reply by omarseleem92** [reply]
@bemcdonnell  Thank you for your quick response. As I understand from swmmio documentation, I can use swmmio to change the variables and save the modified models separately then I can run the models in parallel using swmmio. 

I used swmmio to change some variables such as the subcatchments areas but I could not change the LID control parameters using swmmio. Do you have any recommendations on how to do this? Is there any way to save the model after changing the LID parameters using pyswmm?

**Reply by bemcdonnell** [reply, marked as answer]
@omarseleem92, I think you're on the right track in this case to use pyswmm (I do love swmmio).  I linked you to that module is swmmio so you could see what the code needs to look like to run the process on the command line (not to use swmmio).  Your approach using multiprocessing has the right idea!  Instead of using multiprocessing, you have to use `queue`, `threading`, and `subprocess`.   Take a look at https://docs.python.org/3/library/queue.html .  There are a few things happening here.  a Queue is just a list of jobs descriptions (just like you have in your simulations list).  Threads can be used then to handle the various jobs in the queue.  The difference with what you need to do here is instead of write your pyswmm code into the `worker` function, you need to simply write code that calls a `subprocess`.  And in that subprocess, run a pyswmm_wrapper.py.  For starters, just take what I put in swmmio (pyswmm_wrapper.py) and use that for your project.  You can pass the simulation arguments to that module using command line arguments.

**Reply by omarseleem92** [reply]
@bemcdonnell  Thank you very much. I followed what you said and it worked

**Reply by bemcdonnell** [reply]
@omarseleem92, if you would be willing and open to help us back, do you mind writing the workflow up into a small guide and we can turn it into an example? This would help many others like yourself

**Reply by omarseleem92** [reply]
@bemcdonnell sure I will write it and share it with you.

**Reply by omarseleem92** [reply]
@bemcdonnell here is a guide for the workflow (https://medium.com/@omar.abdwahab92/parallel-simulations-and-green-infrastructure-analysis-with-pyswmm-dbf79409b73e)

**Reply by bemcdonnell** [reply]
@omarseleem92 this is so helpful! Great work.  Would you be open to sharing the files? I can bundle them and put your work onto the pyswmm website examples so that others can benefit. If so, zip them up and drop them in here :-)

**Reply by omarseleem92** [reply]
Yes, sure. Please feel free to use the files. 

I added them also to here (https://github.com/omarseleem92/PySWMM/tree/main/Parallelizing%20PySWMM) 

[Parallelizing PySWMM - Example.zip](https://github.com/pyswmm/pyswmm/files/12240584/Parallelizing.PySWMM.-.Example.zip)

**Reply by omarseleem92** [reply]
Hi @bemcdonnell ,

I'm currently working on extracting the time steps and the corresponding total inflow for an outfall in parallel simulations, following your recommended approach. My strategy was to utilize a manager for a shared list to store the results.



import multiprocessing


```
# Create a manager for a shared list to store results
manager = multiprocessing.Manager()
shared_results = manager.list()
```


In my pyswmm_wrapper.py script, I've set up a dictionary to capture parameters, time steps, and outflow values:


```
 # Initialize a dictionary to store parameters, time steps, and outflow values
    result_data = {
    'A2': args.A2,
    'Surface_thickness_value': args.Surface_thickness_value,
    'Soil_thickness_value': args.Soil_thickness_value,
    'Storage_thickness_value': args.Storage_thickness_value,
    'Drain_offset_value': args.Drain_offset_value,
    'Outflow_value': args.Outflow_value,
    'Seepage_rate_value': args.Seepage_rate_value,
    'time_steps': [],
    'outflow_values': []
    }
```
    
    
During the simulation, I append the current timestep and the corresponding inflow to the lists:
    
     for ind,step in enumerate(sim):
            # append the current timestep to the list
            result_data['time_steps'].append(sim.current_time)
             
            # Append the corresponding outflow to the list
            result_data['outflow_values'].append(Outfall.total_inflow)
            pass


 After the simulation, I append the result_data dictionary to shared_results:

           
```
        # After the simulation, you can append the result_data dictionary to the shared_results
        shared_results.append(result_data)
```   

However, It is not working. I was wondering if you have suggestions on how to do this.
    
Best regards.
Omar

**Reply by bemcdonnell** [reply]
@omarseleem92 i have a feeling that the method in which you are appending is not during a unique state in a multiprocess. Can you try initializing a list of a known length number of simulations and setting an explicit index of your shared results rather than append? I and still not convinced this will work either but worth a try. I would also suggest that you consider writing a small temp file out and simply return the file info (such as a UUID generated file name).  Again this is where threading gets sophisticated. Give those a shot and if they don’t work, we will keep riffing on it :-)
